token/program-2022/src/extension.rs

```diff
@@ -1,13 +1,279 @@
+        Ok((account_type_index, tlv_start_index))
+    }
+}
+
+fn type_and_tlv_indices<S: BaseState>(rest_input: &[u8]) -> Result<(usize, usize), ProgramError> {
+    if rest_input.is_empty() {
+        Ok((0, 0))
+    } else {
+        let type_index = Account::LEN.saturating_sub(S::LEN);
+        let tlv_start_index = type_index.saturating_add(size_of::<AccountType>());
+        Ok((type_index, tlv_start_index))
+    }
+}
+
+/// Encapsulates immutable base state data (mint or account) with possible extensions
+#[derive(Debug, PartialEq)]
+pub struct StateWithExtensions<'data, S: BaseState> {
+    /// Unpacked base data
+    pub base: S,
+    /// Unpacked base data
+    pub account_type: AccountType,
+    /// Slice of data containing all TLV data, deserialized on demand
+    tlv_data: &'data [u8],
+}
+impl<'data, S: BaseState> StateWithExtensions<'data, S> {
+    /// Unpack base state, leaving the extension data as a slice
+    pub fn unpack(input: &'data [u8]) -> Result<Self, ProgramError> {
+        check_not_multisig(input)?;
+        let (base_data, rest) = input.split_at(S::LEN);
+        let base = S::unpack(base_data)?;
+        let (type_index, tlv_start_index) = type_and_tlv_indices::<S>(rest)?;
+        let account_type = AccountType::try_from(rest[type_index])
+            .map_err(|_| ProgramError::InvalidAccountData)?;
+        check_account_type::<S>(account_type)?;
+        Ok(Self {
+            base,
+            account_type,
+            tlv_data: &rest[tlv_start_index..],
+        })
+    }
+
+    /// Unpack a portion of the TLV data as the desired type
+    pub fn get_extension<V: Extension>(&self) -> Result<&V, ProgramError> {
+        if V::ACCOUNT_TYPE != S::ACCOUNT_TYPE {
+            return Err(ProgramError::InvalidAccountData);
+        }
+        let TlvIndices(_, length_start_index, value_start_index) =
+            get_extension_indices::<V>(self.tlv_data, false)?;
+        let length =
+            pod_from_bytes::<Length>(&self.tlv_data[length_start_index..value_start_index])?;
+        let value_end_index = value_start_index.saturating_add(usize::from(*length));
+        pod_from_bytes::<V>(&self.tlv_data[value_start_index..value_end_index])
+    }
+}
+
+/// Encapsulates mutable base state data (mint or account) with possible extensions
+#[derive(Debug, PartialEq)]
+pub struct StateWithExtensionsMut<'data, S: BaseState> {
+    /// Unpacked base data
+    pub base: S,
+    /// Raw base data
+    base_data: &'data mut [u8],
+    /// Writable account type
+    account_type: &'data mut [u8],
+    /// Slice of data containing all TLV data, deserialized on demand
+    tlv_data: &'data mut [u8],
+}
+impl<'data, S: BaseState> StateWithExtensionsMut<'data, S> {
+    /// Unpack the base state portion of the buffer, leaving the extension data as
+    /// a serialized slice.
+    pub fn unpack(input: &'data mut [u8]) -> Result<Self, ProgramError> {
```

Comment:
I suspect there'll be cases in the implementation when one starts with a `StateWithExtensions` and then wants to promote it into a `StateWithExtensionsMut`. I ran into this need with zk-token, which spawned [PodAccountInfoData::into_mut](https://github.com/solana-labs/spl-zk-token/blob/e54c8986d9b8d37b4cc1ddc05846bdffc96e50c4/program/src/pod.rs#L108).

I don't think we need to address this here/now, just pointing this out.  One option could be to break up "unpack()"ing the base state from "new()", so that I could just `StateWithExtensionsMut::new()` from an account I've already used ``StateWithExtensions` with, and then jump into calling `StateWithExtensionsMut::get_extension()` without incurring the overhead of unpacking the base state again

Reply:
Ah ok, my thinking was that `StateWithExtensionsMut` would only be used if you knew that you might change the data later.  The ZK helpers are really great, but since we won't be modifying the base data (mint / account) in-place, it'll be pretty awkward to slot them in.  Are you subtly suggesting moving to in-place modifications for mints / tokens?  That could certainly take down the amount of compute units by the program!